firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/datagrams.txt

```diff
@@ -0,0 +1,446 @@
+    WHEN("A body with a complete 2-byte header, and a non-empty paylaod buffer inconsistent with the length field of the header")
+      THEN("The transform method reports invalid length status")
+      THEN("After the transform method is called, the value returned by the output_datagram's seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the output_datagram's length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The value returned by the output_datagram's length accessor method is inconsistent with the actual length of the payload buffer from the body")
+      THEN("The paylaod returned by the output_datagram's paylaod accessor method is equal to the paylaod from the input_buffer body")
+      THEN("The output_datagram's payload accessor method returns a reference to the payload given in its constructor")
+      THEN("The data in the output_datagram's payload buffer is independent of the data in input_buffer")
+
+    WHEN("A body with a non-empty paylaod buffer and a complete 2-byte header where the sequence field of the header does not match sequence given in output_datagram's constructor")
+      THEN("The transform method reports invalid sequence status")
+      THEN("After the transform method is called, the value returned by the output_datagram's seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the output_datagram's length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The paylaod returned by the output_datagram's paylaod accessor method is equal to the paylaod from the input_buffer body")
+      THEN("The output_datagram's payload accessor method returns a reference to the payload given in its constructor")
+      THEN("The data in the output_datagram's payload buffer is independent of the data in input_buffer")
+
+    WHEN("A body with a complete 2-byte header, a non-empty payload buffer consistent with the length field of the header is given to the receiver")
+      THEN("The transform method reports ok status")
+      THEN("After the transform method is called, the value returned by the output_datagram's seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the output_datagram's length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The paylaod returned by the output_datagram's paylaod accessor method is equal to the paylaod from the input_buffer body")
+      THEN("The output_datagram's payload accessor method returns a reference to the payload given in its constructor")
+      THEN("The data in the output_datagram's payload buffer is independent of the data in input_buffer")
+
+Scenario: Datagram Receiver correctly reports incrementing, rollover, and resetting of expected sequence number
```

Comment:
For certain THENs which are already fully covered by other scenarios, it may be fine to remove those THENs from this scenario if they're irrelevant to the test, e.g. THEN("The data in the output_datagram's payload buffer is independent of the data in input_buffer") and THEN("The output_datagram's payload accessor method returns a reference to the payload given in its constructor") and THEN("The value returned by the output_datagram's length accessor method is equal to the length field of the input_buffer body's header"). Removing some or all of these THENs could make it easier to understand what these tests care about and what their intent is, right now it might be a bit too distracting to keep all of them in this scenario. I think this is a judgement call for you to make.

Reply:
From the test guidelines: "THENs should achieve full coverage of all relevant outputs, state changes, and side-effects for each WHEN." - so it comes down to deciding which of these things are relevant to each scenario. For scenarios covering basic functionality (which I think was all of the scenarios in CRCElement), we might want to cover everything; and for scenarios testing specific behaviors, we should cover some smaller subset of consequences.

Reply:
For this scenario, you could consider using Catch2 data generators for higher coverage of possible values of sequence numbers; up to you.

Reply:
> For certain THENs which are already fully covered by other scenarios, it may be fine to remove those THENs from this scenario if they're irrelevant to the test, e.g. THEN("The data in the output_datagram's payload buffer is independent of the data in input_buffer") and THEN("The output_datagram's payload accessor method returns a reference to the payload given in its constructor") and THEN("The value returned by the output_datagram's length accessor method is equal to the length field of the input_buffer body's header"). Removing some or all of these THENs could make it easier to understand what these tests care about and what their intent is, right now it might be a bit too distracting to keep all of them in this scenario. I think this is a judgement call for you to make.

Yes I concur, I will remove irrelevant cases.

Reply:
> For this scenario, you could consider using Catch2 data generators for higher coverage of possible values of sequence numbers; up to you.

I'm going to do this in a future PR, as generating buffers with sequences ranging from say 0 to 255 is not the issue, but passing that generated buffer in the next iteration is slightly complicated. I will try this later.

Reply:
Sounds good; for certain tests it might also be feasible to just use for loops - and that might be clearer and more readable

Reply:
Yep added.